### DISCUSSION TITLE
Why is it not possible to support Beken chips?

### DISCUSSION THREAD

**Discussion by MatthiasNeuhaus** [arendst/Tasmota#20590] [Ideas]
I just have some minor experience with tasmota as I only did one project for a smart meter with an ESP32-POE. So please be patient with me if this idea is stupid. 

I have a "smart" led controller and was afraid if I can also flash it with tasmota. After some googling I found out that the build in WB3S chip is not supported by tasmota (and would need to be replaced by a esp) but by https://github.com/openshwprojects/OpenBK7231T_App. 

Soldering is an option for me for sure, but I don't know jet if I have any drawbacks from just using OpenBK. I would like build up a smart home and keep the connection first just on tasmota level, so I'm independent of an home assistant installation. 

If there was somebody able to get the chip running in the above project, why is it not possible to support it by tasmota? Is it "just" lag of manpower, some philosophy thing or architektural Problems? As the other project is also open source, it should be possible to spike the ideas to support the chipset?! 

To sum up, the idea is to integrate Beken chip support into tasmota, as they seem to kind of displace the esp chips in some areas?!

**Comment by sfromis** [comment]
https://github.com/arendst/Tasmota/discussions/12022

Asking "possible" is not very useful, when it is about major amounts of work to have both Arduino Core and lots of libraries broadly working on BK7231T. If that was already in place, adapting Tasmota would become feasible.

Hence a comparison with a simpler project (with only a part of Tasmota features), and starting on the basis of the Beken chips is no illustration of it being "reasonably feasible".

The drawbacks from using OpenBK7231T is that it does not have the features of Tasmota. That does not make it unusable for many use cases, where it can be perfectly workable. However, since you talk about "independent of an home assistant installation", it does sound like you may be aiming for more, as "Home Assistant compatibility" is a key "selling point" for the Beken-based project, but IDK.

**Comment by MatthiasNeuhaus** [comment]
Ok when I correctly understand (https://github.com/arendst/Tasmota/discussions/12022) it is an architectural reason. Arduino IDE - that tasmota is based on - does not support Beken chips. Fair enough tasmota is not able to support them without huge effort. That is answering my question and making the Idea useless. 

May be possible that openBK does not have all features, but as far as I read/ understand I can directly connect it with tasmota devices via MQTT: 
"Thanks to keeping Tasmota standard, OBK has basic compatibility with ioBroker and similar systems through TELE/STAT/CMND MQTT packets, Tasmota Control app is also supported" 
...and that all I need right now. So I'll give it a try. If it is not working, I'll solder an ESP-12S on it that I already ordered 😉

**Reply by sfromis** [reply]
Well, strictly speaking it is a long time since Tasmota even supported using Arduino IDE (PlatformIO and possibly VSCode is used instead), but Arduino Core (which is not part of the IDE) is providing a lot of functions to Tasmota.

While I've not used OpenBK7231T_App, my understanding is that it, for supported devices, can work well for having them connected via MQTT.

IIRC, I saw some barebones attempt at porting Arduino Core to the Beken chips, but not at a level needed for Tasmota, and its a long time since I've seen anything there.

**Reply by MatthiasNeuhaus** [reply]
@sfromis, thanks a lot for your answers!

**Reply by brjhaverkamp** [reply]
I was looking for the same. I know Tasmota only supports the ESP family of processors atm. But it would be great if it can be generalized to be able to support other platforms. I understand that will require a lot of work. And some sort of abstraction layer towards the hardware.
I guess it is similar to the time Linux moved from X86 to other processor types or like openwrt supporting many different platforms. But ESPs will one day be overtaken by other chips or there will be many different ones. 
I hope it can start with the beken chips, now that there is something similar to tasmota for that one. Maybe it would be possible to merge it somehow and have a shared platform/codebase  

Regards,

Bert

**Comment by probonopd** [comment]
https://docs.libretiny.eu/

> PlatformIO development platform for BK7231 and RTL8710 IoT chips.
> 
> The main goal of this project is to provide a usable build environment for IoT developers. While also providing vendor SDKs as PlatformIO cores, the project focuses on developing working Arduino-compatible cores for supported families. The cores are inspired by Espressif's official core for ESP32, which should make it easier to port/run existing ESP apps on less-common, unsupported IoT modules.

So maybe it will become more feasible, given this?

**Reply by sfromis** [reply]
Why do you think so? 

What I mentioned earlier was
> when it is about major amounts of work to have both Arduino Core and lots of libraries broadly working on BK7231T. If that was already in place, adapting Tasmota would become feasible.

Sure, in principle it would "not be impossible" to rewrite many parts of Tasmota to no longer rely on functions from Arduino Core (and many libraries), it's just that it would be a large project to pull through. Nothing stopping "someone" from making a fork, and then do a lot of coding work. Who?

**Comment by Jason2866** [comment]
The core development Team of Tasmota is simply to small to support more than one chip manufacturer
